[Hosting Dashboard] monetize section

millerf opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Fixes https://linear.app/a8c/issue/CHE-243/hosting-dashboard-migrate-membershipsdonations-purchases-list-to

## Proposed Changes

* Add the monetize section to the Hosting Dashboard
* Modified FlashMessage to work properly using a `param` prop instead of protected `key`
* Views:

| Subscriptions | Subscription detail |
| --- | --- |
| <img width="1038" height="446" alt="Screenshot 2025-09-29 at 18 31 17" src="https://github.com/user-attachments/assets/ae89b9c4-35b1-49df-978c-d4ebc85b6db8" /> | <img width="789" height="661" alt="Screenshot 2025-09-29 at 18 27 57" src="https://github.com/user-attachments/assets/1ac14408-e7cb-48a9-809d-5560c0ca98e9" /> |




## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run `ENTRY_LIMIT=entry-dashboard-dotcom,entry-dashboard-a4a yarn start` or use the calypso live link provided
* Go to http://calypso.localhost:3000/v2/me/billing/monetize-subscriptions
* Make sure the experience matches the tavle at the bottom of the https://wordpress.com/me/purchases page (if you have subcribed to any memberships).
* Otherwise switch to Sandbox Store and subskcribe to https://bestdotcroissants.wordpress.com/2024/12/30/294/ with the `4242424242424` card

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

millerf committed:
Added Monetize Subscriptions route

millerf committed:
Added link to main menu header

millerf committed:
First import

millerf committed:
Added data layer

millerf committed:
Added memberships subscriptions fetcher and remove use of useBreakpoints

millerf committed:
Fix typings

millerf committed:
Some cleanup, remove unused code (MembershipSite) and remove usage of moment.js

millerf committed:
Created Monetize subscription route

fushar commented on client/dashboard/me/me-menu/index.tsx:
```diff
@@ -14,6 +14,9 @@ const MeMenu = () => {
 			<ResponsiveMenu.Item to="/me/profile">{ __( 'Profile' ) }</ResponsiveMenu.Item>
 			<ResponsiveMenu.Item to="/me/billing">{ __( 'Billing' ) }</ResponsiveMenu.Item>
+			<ResponsiveMenu.Item to="/me/billing/monetize-subscriptions">
+				{ __( 'Monetize Subscriptions' ) }
```
Do we need this? The other screens under /me/billing are not getting added as a submenu here 🤔

millerf replied:
We are not sure wether to add this screen in the billing menu or directly in the main header. So I put it in both places until a decision is made.

fushar commented on packages/api-core/src/monetize/mutators.ts:
```diff
@@ -0,0 +1,9 @@
+import { MembershipSubscription } from './types';
+
+export function normalizeMonetizeSubscription(
```
Looks like this function is doing nothing; I think we can just remove it?

millerf replied:
Still working on it...

fushar commented on packages/api-core/src/monetize/types.ts:
```diff
@@ -0,0 +1,21 @@
+export interface MembershipSubscription {
```
Should this be inside `/me-monetize` so that it's adjacent to the fetcher?

millerf replied:
I followed the approch taken by [purchases](https://github.com/Automattic/wp-calypso/tree/873d07f8733e9027ee253fe4dffd9e8d80c1690b/packages/api-core/src/purchase). Happy to move it...

fushar commented on packages/api-core/src/me-monetize/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './fetchers';
```
I think the directory should be called `/me-memberships` to align with the API name. See: pgz0xU-df-p2

millerf replied:
I wanted to take the opportunity to update the name: the feature is now called Monetize, although the API is still under memberships...

fushar commented on client/dashboard/me/monetize-subscriptions/dataviews.tsx:
```diff
@@ -0,0 +1,109 @@
+			enableSorting: false,
+			enableHiding: false,
+			filterBy: false,
+			getValue: ( { item }: { item: MembershipSubscription } ) => {
```
[nit] I believe `: { item: MembershipSubscription }` is not needed / can be implied, as the return type is already marked as `Fields< MembershipSubscription >`.

millerf replied:
The linter does not let me remove it ;)

fushar commented on client/dashboard/me/monetize-subscriptions/dataviews.tsx:
```diff
@@ -0,0 +1,109 @@
+			},
+			render: ( { item }: { item: MembershipSubscription } ) => {
+				return (
+					<div className="membership-item__information purchase-item__information">
```
Do we need all these classes? Seems like there's no CSS styles applied to them.

millerf replied:
Yes, nice catch

fushar commented on client/dashboard/me/monetize-subscriptions/membership-item.tsx:
```diff
@@ -0,0 +1,94 @@
+
+	useEffect( () => {
+		async function fetchData() {
+			const data = await fetch( 'https://public-api.wordpress.com/rest/v1.1/sites/' + siteId );
```
We should use our queries; I think this is `siteByIdQuery()`?

millerf replied:
Oh. I'll try that

millerf replied:
Done!

millerf committed:
use siteByIdQuery

millerf committed:
Remove unecessary classes

millerf committed:
Revert unecessary change

millerf committed:
Make subscription detail route

millerf committed:
Start work on Monetize subscription details

millerf committed:
More work on Subscription details

millerf committed:
Ignore CLAUDE settings

millerf committed:
Subscription details almost working

millerf committed:
Create useNotice

millerf committed:
Added some mutations

millerf committed:
Rmeove unecessary file

millerf committed:
Some fixes

millerf committed:
Button

millerf committed:
Some refactor

millerf committed:
Enabling/Disabling auto-renew works

millerf committed:
Refactor style for subscription details

millerf committed:
Fix lint

arthur791004 commented on client/dashboard/app/hooks/use-notice.ts:
````diff
@@ -0,0 +1,18 @@
+ * ```
+ */
+
+export const useNotice = () => {
````
We discussed on another PR and we have decision to remove it 🙂

millerf replied:
Oh. I'll revert all this... Thanks for the heads up.

millerf replied:
Done in 60c001d14ea73d28b4e14fa1bf7233a89666864e

arthur791004 commented on client/dashboard/app/router/me.tsx:
```diff
@@ -198,6 +197,28 @@ export const paymentMethodsRoute = createRoute( {
 	)
 );
 
+export const monetizeSubscriptionsRoute = createRoute( {
```
Could you help to add `head` property to set the `document.title` for this screen as above route did.

millerf replied:
Sure!
Done in 6df56b405560b2006bd8563e75f57dea12a71eac

fushar commented on packages/api-core/src/monetize/fetchers.ts:
```diff
@@ -0,0 +1,10 @@
+import { wpcom } from '../wpcom-fetcher';
+import { MembershipSubscription } from './types';
+
+export async function fetchMonetizeSubscriptions(): Promise< MembershipSubscription[] > {
```
The file should ideally be called `api-core/me-memberships.ts`. See: pgz0xU-df-p2

millerf replied:
I moved it to `api-core/src/me-monetize`

fushar commented on .gitignore:
```diff
@@ -133,3 +133,6 @@ packages/plans-grid-next/static/mockServiceWorker.js
 playwright-report/
 /cookies
 output/
+
```
I don't think this change belongs to this PR :)

millerf replied:
Well, it could as well be,...

millerf replied:
Created https://github.com/Automattic/wp-calypso/pull/105901

millerf committed:
Remove useNotice

millerf committed:
Added pages titles

millerf committed:
Rename all Memberships instances

millerf committed:
Rename monetize into me-memberships

millerf committed:
Show a flash message when the subscription is deleted

p-jackson commented on packages/api-queries/src/me-monetize.ts:
```diff
@@ -0,0 +1,75 @@
+	queryOptions( {
+		...monetizeSubscriptionsQuery(),
+		enabled: !! subscriptionId,
+		select: ( data ) => {
```
Beware: by putting the `select` function inside `api-queries` it makes the query a little less reusable.

When you pass a `monetizeSubscriptionQuery` to the `useQuery()` or `useSuspenseQuery()` hooks, then those hooks will return a single `MonetizeSubscription` as expected. Because hooks run the `select` function.

However, when you pass a `monetizeSubscriptionQuery` to the `queryClient.ensureQueryData()` or `queryClient.getQueryData()` functions, these functions will still return a whole `MonetizeSubscription[]` array.

The `select` functions are there to help minimise renders, not necessarily map the data types like this. But it is a pattern that gets used sometimes. I just wanted to make sure you were aware of the subtlety.

millerf replied:
Thanks for this clarification. I think it is still fine. However let me know if there is a more canonical way to do what I need. Meaning:
- `monetizeSubscriptionsQuery` returns the whole list of subscriptions
- `monetizeSubscriptionQuery` returns a specific subscription

I guess what I could do is in `monetizeSubscriptionQuery`, call `queryClient.getQueryData( monetizeSubscriptionsQuery. key)` and filter the result.
I'll try this new approach if you think it makes more sense.

p-jackson replied:
> let me know if there is a more canonical way to do what I need

I'm not sure we have come up with a more canonical way tbh. Building one query on top of another one (like building `monetizeSubscriptionQuery` on top of `monetizeSubscriptionsQuery`) _usually_ works out ok. Because _usually_ when we call `ensureQueryData()` we're just preloading in the router, and so we don't really care what the query is returning.

millerf replied:
I see!

fushar replied:
We now have "official" guideline for this :)


https://github.com/Automattic/wp-calypso/blob/054e228306e2a907b9caed7b3b2cdf4460ce17f6/packages/api-queries/README.md?plain=1#L27

In this situation, I think the best way is to add the filter in the `queryFn()` instead, e.g.:

```ts
export const monetizeSubscriptionQuery = ( subscriptionId: string ) =>
	queryOptions( {
		...monetizeSubscriptionsQuery(),
		queryFn: () => {
				const subscriptions = await fetchMonetizeSubscriptions();
				const subscription = data?.find( ( sub ) => sub.ID === subscriptionId );
				if ( ! subscription ) {
					throw new Error( `Subscription with ID ${ subscriptionId } not found` );
				}
				return subscription;
		},
}
```

This way we get all the benefits:

- We're reusing the same query cache as `monetizeSubscriptionsQuery()`
- If by any chance some component use `queryClient.ensureQueryData()`, it will return the single subscription correctly.

What do you think? Also cc @p-jackson

p-jackson replied:
> We're reusing the same query cache as `monetizeSubscriptionsQuery()`

That's a problem though isn't it? The `monitizeSubscriptionsQuery()` will write `MonetizeSubscription[]` to the cache. Then `monetizeSubscriptionQuery()` comes along and overrides it with a `MonetizeSubscription` which is a whole different data type.

`monetizeSubscriptionQuery()` really needs a query key that includes `subscriptionId`.

fushar replied:
Wait sorry you're right my bad, I got confused 🤦 

Yeah, we need a different query key. But we need to make sure we're invalidating / setting the query data of the `monitizeSubscriptionsQuery()` when a particular `subscriptionId` is changed, which I think is already being done here in this PR.

millerf replied:
> which I think is already being done here in this PR.
Yes, it is.

Am I understanding it well that the solution in this PR follows the new rules as well. Meaning there is no change needed?

fushar replied:
I think we need to do something like this:

```ts
export const monetizeSubscriptionQuery = ( subscriptionId: string ) =>
	queryOptions( {
		[ 'me', 'monetize', 'subscriptions', subscriptionId ],
		queryFn: () => {
				const subscriptions = await fetchMonetizeSubscriptions();
				const subscription = data?.find( ( sub ) => sub.ID === subscriptionId );
				if ( ! subscription ) {
					throw new Error( `Subscription with ID ${ subscriptionId } not found` );
				}
				return subscription;
		},
}
```

millerf replied:
@fushar This is doing an unecessary query `fetchMonetizeSubscriptions()` for each subscription. My solution was caching the list of subscriptions and using said cache.

millerf replied:
I used:
```
export const monetizeSubscriptionQuery = ( subscriptionId: string ) =>
	queryOptions( {
		queryKey: [ ...MonetizeQueryKeys, subscriptionId ],
		queryFn: async () => {
			const subscriptions: MonetizeSubscription[] = await queryClient.ensureQueryData( {
				queryKey: monetizeSubscriptionsQuery().queryKey,
			} );
			const subscription = ( subscriptions ?? [] ).find( ( sub ) => sub.ID === subscriptionId );
			if ( ! subscription ) {
				throw new Error( `Subscription with ID ${ subscriptionId } not found` );
			}
			return subscription;
		},
	} );
```

To fix that issue

p-jackson replied:
This is a novel approach 😅 This doesn't break any rules about React Query; React Query can technically monitor any sort of async process, so there's no reason it can't monitor a call to `ensureQueryData`.

For me though, using `select` is the idiomatic way of grabbing a small bit out of data out of an existing query. The only controversial point is whether that `select` function should be in the `api-queries` layer or the component layer. If `select` is defined in `api-queries` then there are some weird nuances as outlined above, but it means we get to have `monetizeSubscriptionQuery` which gets to present an abstraction as if there is some endpoint that returns individual subscriptions. If `select` is done in the component then `api-queries` will only expose `monetizeSubscriptionsQuery` which returns multiple `MonetizeSubscription`s. The query layer won't be hiding the fact that there's no way to request one at a time.

I slightly lean towards saying that `select` in the component layer is better. That is, deleting `monetizeSubscriptionQuery`, and using `monetizeSubscriptionsQuery` in the component with a `select` option. `monetizeSubscriptionQuery` is only used in one place, so it's not like the abstraction of pretending that there is an endpoint that will return us a single subscription is saving us much duplication.

But I think this is going to be a relatively common thing (having a endpoint that returns a list where the component only wants one of them). I don't think this sort of `queryFn` is what we should be doing for all of these cases. We should just use `select`.

fushar replied:
I see I see. OK let's not block this PR over this. We can discuss about this pattern outside of this PR :)

But it's great that we all are learning something!

p-jackson commented on packages/api-queries/src/me-monetize.ts:
```diff
@@ -0,0 +1,75 @@
+		},
+	} );
+
+const UpdateSubscriptionCache =
```
nitpick: I'd expect a lowercase at the start

```suggestion
const updateSubscriptionCache =
```

millerf replied:
Done!

p-jackson commented on packages/api-queries/src/me-monetize.ts:
```diff
@@ -0,0 +1,75 @@
+			 * used to display the notification labeled "This item has been removed".
+			 */
+			if ( response.redirect ) {
+				window.location.assign( response.redirect );
```
This redirecting doesn't look like it should  be in the `api-queries` package. Would it make it hard to use this query in other contexts?

I see the redirect URL comes from the server, so we're being instructed to redirect from something outside the UI layer 🤔 Maybe the `monetizeSubscriptionStop` query should accept a `followRedirects` prop at least?

millerf replied:
Mmh. This redirect is used on Jetpack sites to remove the cookie and remove access to some specific content.
this is very needed. The user is then redirected back to calypso.

p-jackson replied:
Interesting. We have similar situations with the password change and account deletion mutations. In both cases we do the reloading in the UI layer instead of in `api-queries`. But in those cases we also aren't returned a redirect URL from the API 🤔 I'm happy to leave this as it is for now. I find it unexpected that the `api-queries` package might redirect me, but perhaps we need more examples of this sort of situation.

millerf replied:
> I find it unexpected that the api-queries package might redirect me, but perhaps we need more examples of this sort of situation.

I understand. If that help, this particular API call is very specific and is not used anywhere else...

p-jackson commented on client/dashboard/components/flash-message/index.tsx:
```diff
@@ -14,7 +14,7 @@ interface FlashMessageProps {
  */
 export default function FlashMessage( {
-	key = 'updated',
+	param = 'updated',
```
Thanks for catching. This seems fine. Maybe `paramName`? I was trying to choose names that avoided the confusion between which prop was referring to the query param value and which was referring to the query param name. And `param` _might_ be interpreted as the param's value 🤷‍♂️

millerf replied:
Happy to change it to `paramName`. I'll amend it now!

millerf replied:
What about "overrideParamName"? This way devs will know there is no need to pass it?

millerf replied:
What do you think of 351105739a0511990c9549952ca4b20ac0ab15e6

p-jackson commented on client/dashboard/me/monetize-subscriptions/index.tsx:
```diff
@@ -0,0 +1,121 @@
+			<div ref={ ref }>
+				{ /* Show a flash message if the URL contains ?showSuccessRemoved=true when a subscription is removed */ }
+				<FlashMessage
+					param="showSuccessRemoved"
```
By having a single `FlashMessage` component, I was hoping that we could start to use a consistent query param name throughout the hosting dashboard. Is there a reason we couldn't use `?updated=product-removed` query param?

Is there a better query name than `updated` that you think we could use instead?

millerf replied:
Oh. I see. I can change that as well. I also thought `updated` was too generic...
Also, I'll add an helper to FlashMessage to add the right parameter to the URL. This way there is no need to hardcode `updated` everywhere.

millerf replied:
See https://github.com/Automattic/wp-calypso/pull/105695#discussion_r2367905630

p-jackson commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,307 @@
+	const isDisabledAutorenewing = isRenewable && ! subscription.renew_interval;
+	const isUpdating = isEnablingAutoRenew || isDisablingAutoRenew;
+	const navigate = useNavigate();
+	useEffect( () => {
```
I haven't looked into it deeply, but using `useEffect` for error handling seems like a code smell. We shouldn't need to use a `useEffect` to listen for the value of `stoppingStatusError` to change, when we can listen to `onError` of the `monetizeSubscriptionStop` mutation e.g.

```ts
stopSubscription( null, {
  onError: () => {
    // error notice
  }
} );
```

millerf replied:
Yes, I agree. I merely reuse the original code. I should not have.

millerf replied:
I removed this useEffect

millerf committed:
Fix case issue

millerf committed:
Some refactor of FlashMessage

- add addParamForFlashMessage helper
- rename key/param to overrideDefaultParam

millerf committed:
Move use of stoppingStatusSuccess

millerf committed:
Remove unecessary useEffect

millerf committed:
Fix lint

millerf committed:
Fix some issue with previous commit

p-jackson commented on packages/api-queries/src/me-monetize.ts:
```diff
@@ -0,0 +1,76 @@
+export const monetizeSubscriptionStop = ( subscriptionId: string ) =>
+	mutationOptions( {
+		mutationFn: () => requestSubscriptionStop( subscriptionId ),
+		onSuccess: ( response: { redirect?: string } ) => {
```
Could we define the response body type on the `requestSubscriptionStop()` function instead. The `mutationFn` should define the return type in the `api-core` package, and then this `onSuccess` handler should pick up the type automatically.

We should also define return types for the `requestAutoRenewDisable()` and `requestAutoRenewResume()` functions too.

millerf replied:
Done in 81f2836c62b5c5c93a1c14118b18bdd6f407a151

p-jackson commented on client/dashboard/components/flash-message/index.tsx:
```diff
@@ -3,18 +3,27 @@ import { store as noticesStore } from '@wordpress/notices';
 	type?: 'success' | 'error';
 }
 
+const DEFAULT_PARAM_NAME = 'show-flash-message';
```
I prefer just `?flash=` to `?show-flash-message=`

p-jackson commented on client/dashboard/components/flash-message/index.tsx:
```diff
@@ -3,18 +3,27 @@ import { store as noticesStore } from '@wordpress/notices';
 
+const DEFAULT_PARAM_NAME = 'show-flash-message';
+
+export function addParamForFlashMessage(
```
I like the idea of having a helper for setting up the flash message. I'm not sure what the function signature should be though. Because in 351105739a0511990c9549952ca4b20ac0ab15e6 since you're using the `navigate()` function you want the query params as an object.

But [here](https://github.com/Automattic/wp-calypso/blob/f6552304109f4d83e83be7b81a1ec118283b90ca/client/dashboard/me/preferences-new-hosting-dashboard/index.tsx#L99) and [here](https://github.com/Automattic/wp-calypso/blob/f6552304109f4d83e83be7b81a1ec118283b90ca/client/dashboard/me/security-password/index.tsx#L48) we want a query string.
Is it possible to change 351105739a0511990c9549952ca4b20ac0ab15e6 so it uses a query string? Does the `navigate()` function still work if the `addParamForFlashMessage()` function worked more like:

```ts
navigate( addParamForFlashMessage( getMonetizeSubscriptionsUrl(), 'true' ) )
```

---

Feel free to ignore this for this PR in particular, but now I'm wondering whether the flash message tools shouldn't refer to query params at all, and perhaps the correct idea is to call it a "message id".

So in your component component you render your message like
```tsx
<FlashMessage id="removed" message="Thing was removed" />
```

And then you add flash message ids to URLs like
```tsx
const url = '/my-dashboard-page';
window.location.href = addFlashMessage( url, 'removed' );` // internally this can just be a call to core's `addQueryArgs()` function
```

millerf replied:
> But [here](https://github.com/Automattic/wp-calypso/blob/f6552304109f4d83e83be7b81a1ec118283b90ca/client/dashboard/me/preferences-new-hosting-dashboard/index.tsx#L99) and [here](https://github.com/Automattic/wp-calypso/blob/f6552304109f4d83e83be7b81a1ec118283b90ca/client/dashboard/me/security-password/index.tsx#L48) we want a query string.

Wouldn't it be better if those places would use `navigate()` instead of window.location?

millerf replied:
I ended up moving to use URL directly. I also used addFlashMessage whenever necessary. See this commit: 555a9990e420de2a0b74c01b47a48ad739aeae22

p-jackson replied:
It looks good 👍

> Wouldn't it be better if those places would use `navigate()` instead of window.location?

I like `window.location` because its clear we're going to do a real browser load, not something related to a JS router library. I would have thought `navigate()` would not do a page refresh? Maybe I don't fully understand how it works 🙂

millerf committed:
Define return types

millerf committed:
Refactor flash message helper

- Rename to addFlashMessage
- Only accepts URL/string
- Rename parameter to Flash Message ID

Initsogar commented on client/dashboard/components/flash-message/index.tsx:
```diff
@@ -36,7 +47,7 @@ export default function FlashMessage( {
 			}
 
-			params.delete( key );
+			params.delete( overrideDefaultParam );
```
Should this be `id` instead of `overrideDefaultParam`?

```suggestion
			params.delete( id );
```

millerf replied:
Nice catch. We had a few back and forth, and I missed this one...

millerf replied:
Done in f8a8008ce725d11eedaa2b79c00da3ed74316396

Initsogar commented on client/dashboard/me/monetize-subscriptions/monetize-item.tsx:
```diff
@@ -0,0 +1,87 @@
+				event.preventDefault();
+				window.location.href = subscription.site_url;
+			} }
+			title={ String(
```
`sprintf()` returns a string, so no additional casting is needed

millerf replied:
Done!

Initsogar commented on client/dashboard/me/monetize-subscriptions/monetize-item.tsx:
```diff
@@ -0,0 +1,87 @@
+				// eslint-disable-next-line @wordpress/i18n-translator-comments
+				sprintf( __( 'Visit %(siteUrl)s' ), {
+					siteUrl: subscription.site_url,
+				} )
```
I wonder if we could add the translator comments instead of disabling the linter, something like:

```suggestion
				sprintf(
					// translators: %(siteUrl)s is the URL of the site
					__( 'Visit %(siteUrl)s' ),
					{
						siteUrl: subscription.site_url,
					}
				)
```

millerf replied:
Done in fade306d07efdcc81a9a1ede26d451112ab2dcb4

Initsogar commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,307 @@
+				</VStack>
+				<VStack>
+					<b>{ name }</b>
+					<div>{ domain }</div>
```
nit: maybe we can use the [Text](https://github.com/Automattic/wp-calypso/tree/trunk/client/dashboard/components/text) component from the HD on these two:

```suggestion
					<Text weight="700">{ name }</Text>
					<Text>{ domain }</Text>
```

Initsogar replied:
It can be imported from:

```
import { Text } from '../../components/text';
```

millerf replied:
Done in 15fdc7c17f9b1a5e91acb9ab4f13ec029d131641

Initsogar commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,307 @@
+						<HStack alignment="center">
+							<VStack>
+								<HStack alignment="center">
+									<h3>{ subscription.title }</h3>
+									<b>
+										{ formatCurrency(
+											parseFloat( subscription.renewal_price ),
+											subscription.currency
+										) }
+									</b>
```
nit: sames as [here](https://github.com/Automattic/wp-calypso/pull/105695/files#r2376503732).


```suggestion
									<Text as="h3">{ subscription.title }</Text>
									<Text weight="700">
										{ formatCurrency(
											parseFloat( subscription.renewal_price ),
											subscription.currency
										) }
									</Text>
```

millerf replied:
Done in 15fdc7c17f9b1a5e91acb9ab4f13ec029d131641. I actually used size="title"...

millerf committed:
Remove missed instance of overrideDefaultParam

millerf committed:
Add translators comments

millerf committed:
Use <Text> as much as possible

millerf committed:
Use as=h3

paulopmt1 commented on client/dashboard/me/monetize-subscriptions/index.tsx:
```diff
@@ -0,0 +1,117 @@
+		() => [
+			{
+				id: 'manage-purchase',
+				label: __( 'Manage purchase' ),
```
Is it worth replacing this with the button itself to avoid two clicks (since we don’t have more actions)?

<img width="1564" height="803" alt="Image" src="https://github.com/user-attachments/assets/2927f5b4-64f5-45ce-bcae-664f409c383f" />

paulopmt1 commented on client/dashboard/me/monetize-subscriptions/monetize-item.tsx:
```diff
@@ -0,0 +1,87 @@
+	subscription: MonetizeSubscription;
+} ) => {
+	if ( subscription.end_date === null ) {
+		return createInterpolateElement( __( 'Purchased from <MonetizeSiteLink/>' ), {
```
Can we use a button with a more visually appealing layout?

<img width="1564" height="803" alt="Image" src="https://github.com/user-attachments/assets/7b1a0c86-7c9e-4cb8-aa31-02819f61b07c" />

fushar replied:
Maybe just use a simple link (`<a>`) here, similar to `/v2/me/billing/purchases`?

<img width="467" height="136" alt="Image" src="https://github.com/user-attachments/assets/4a50bfc8-77fa-4c74-b22c-e91ea042b5ad" />

paulopmt1 commented on client/dashboard/me/monetize-subscriptions/dataviews.tsx:
```diff
@@ -0,0 +1,99 @@
+			console.error( 'Cannot display manage purchase page for subscription without ID' );
+			return;
+		}
+		return `/me/purchases/other/${ subscriptionId }`;
```
Is this link to the old experience intentional?

<img width="1527" height="575" alt="Image" src="https://github.com/user-attachments/assets/b13e4101-1269-4b1e-b30c-9f9abf370f6a" />

millerf replied:
Oh. Nice catch

millerf replied:
Fixed in 91aacc0ae276b0af23261cce7c92f6fdb1f90d15

paulopmt1 replied:
Nice, thanks for fixing it!

paulopmt1 commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,310 @@
+							/>
+						) }
+
+						<StopSubscriptionButton
```
Do we need to repeat that copy 3 times? What about only showing the CTA Button instead?

<img width="768" height="545" alt="Image" src="https://github.com/user-attachments/assets/07eb434f-e927-408b-bf81-ac1a32f7d0eb" />

paulopmt1 replied:
Also on the removal experience, I see some issues:
- We're not asking if the user really wants to do that (which is a pattern we expect to see on HD. See p1755089977975119-slack-C0982082MSR). 
- Once I click on Remove, I'd expect the button to be blocked and have a loader on it
- Showing a notice on the top with the "Removing this product" doesn't feel the right pattern, especially since it makes the screen jump:

https://github.com/user-attachments/assets/05f0df39-f477-4e21-a3ab-51b8d7c80a64

paulopmt1 commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,310 @@
+									<Text weight={ 700 } size="title" as="h3">
+										{ subscription.title }
+									</Text>
+									<Text weight={ 700 }>
```
This price looks a bit off, should it have the same size of the left text?

<img width="793" height="536" alt="Image" src="https://github.com/user-attachments/assets/bc39b591-02c7-449b-9e23-b4f66518146e" />

millerf replied:
I re-done all this view with components from the Billing views

paulopmt1 reviewed (commented):
Thanks for working on this @millerf. 
I saw you have code feedback, so I focused on UX. They're not blockers for this PR, so you can do a follow-up. 

I'd also recommend checking them with designers to see how we can better address that feedback.

millerf committed:
Fix old style vs new style URLs

millerf committed:
Added autorenew management

millerf committed:
Refactor view with PurchaseSettingsCard

fushar commented on packages/api-queries/src/index.ts:
```diff
@@ -30,6 +30,7 @@ export * from './me-payment-methods';
 export * from './me-preferences';
 export * from './me-preferences-login';
 export * from './me-purchases';
+export * from './me-monetize';
```
[nit] Please help sort the exports :)

millerf replied:
Done in 481cf4885a7b778efa03dade20693344e69fd2bf

fushar commented on packages/api-queries/src/me-monetize.ts:
```diff
@@ -0,0 +1,78 @@
+export const monetizeSubscriptionQuery = ( subscriptionId: string ) =>
+	queryOptions( {
+		...monetizeSubscriptionsQuery(),
+		enabled: !! subscriptionId,
```
We now have an "official" guideline for this:

https://github.com/Automattic/wp-calypso/blob/054e228306e2a907b9caed7b3b2cdf4460ce17f6/packages/api-queries/README.md?plain=1#L49

We should let the component to add `enabled` when necessary. Also currently, with this inline logic here, it would seems that `subscriptionId` is sometimes unavailable, but the type is `string` (not `?: string`). It's better to assume that this function will always accept a defined `subscriptionId`, and let the component itself pass `enabled` when not.

millerf replied:
Done in 59a077d8f3f87354d491da107920f396a9e0b090
In the end, `subscriptionId` has to be a non-empty string. So I removed the `enabled`.

A question however: the right way woud be to do something like this in the component?

```
	const { data: subscription } = useQuery( {
               ...monetizeSubscriptionQuery( subscriptionId ),
               enabled: someCondition,
        });

```

OR you were suggesting to pass it as a parameter in `monetizeSubscriptionQuery`?

fushar replied:
> A question however: the right way woud be to do something like this in the component?

Yes!

p-jackson commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,364 @@
+	const { createErrorNotice } = useDispatch( noticesStore );
+	const navigate = useNavigate();
+	const title = isProduct
+		? // eslint-disable-next-line @wordpress/i18n-translator-comments
```
Since we have to write a comment to ignore the eslint warning anyway, why not just write a comment to provide the translator note? It's always good to provide them with an idea of what sort of "thing" `%s` will be so they have an idea about gender etc.

millerf replied:
Done in 5e2eb65dd5654eae5609247f0f83f37a8d0c043d

p-jackson commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,364 @@
+	return (
+		<ActionList.ActionItem
+			title={ title }
+			description={ __( "We'll be sorry to see you go!" ) }
```
```suggestion
			description={ __( "We’ll be sorry to see you go!" ) }
```

p-jackson commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,364 @@
+	const title = isAutoRenewing ? __( 'Disable auto-renew' ) : __( 'Enable auto-renew' );
+	const onError = () => {
+		if ( isProduct ) {
+			createErrorNotice( __( 'There was a problem while removing your product.' ), {
```
Please update the snackbars to conform to the style guide here
https://github.com/Automattic/wp-calypso/blob/4f53b1b6fb44c1675d3c6383142608ba9ef8569c/client/dashboard/docs/typography-and-copy.md#L21

millerf replied:
Done. Changed the copy.

p-jackson commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,364 @@
+		>
+			{ isStoppingSubscription && (
+				<Notice status="info">
+					{ isProduct ? __( 'Removing this product' ) : __( 'Stopping this subscription' ) }
```
These notice messages should end in periods.

```suggestion
					{ isProduct ? __( 'Removing this product.' ) : __( 'Stopping this subscription.' ) }
```

millerf replied:
Done in f48060f280e89f4b4d29a63173ce9a050b51fb88

p-jackson replied:
I think this updated `client/me/memberships/subscription.jsx` but not `client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx`

p-jackson commented on client/dashboard/me/monetize-subscriptions/monetize-item.tsx:
```diff
@@ -0,0 +1,87 @@
+		return <SiteIcon site={ site } size={ 36 } />;
+	}
+
+	return <></>;
```
Would it be more idiomatic to return `null`?

millerf replied:
Done in 8e04b54e1aa26d744435cc0e8ef345596610cfef

p-jackson commented on client/dashboard/me/monetize-subscriptions/dataviews.tsx:
```diff
@@ -0,0 +1,91 @@
+			enableSorting: false,
+			enableHiding: false,
+			filterBy: false,
+			getValue: ( { item }: { item: MonetizeSubscription } ) => {
+				return item.site_id + ' ' + item.site_title + ' ' + item.site_url;
+			},
+			// Render the site icon
+			render: ( { item }: { item: MonetizeSubscription } ) => {
```
In cases like this where TS knows the type already I think its better to leave it out. Seems its more resilient to if the dataviews package changes the API or something.

```suggestion
			getValue: ( { item } ) => {
				return item.site_id + ' ' + item.site_title + ' ' + item.site_url;
			},
			// Render the site icon
			render: ( { item } ) => {
```

p-jackson reviewed (commented):
Thank you for iterating on `<FlashMessage>`! I like the new `addFlashMessage()` function.

I left some code feedback that would be good to address, but otherwise I think this is looking good from a code point of view.

p-jackson commented on client/dashboard/components/flash-message/index.tsx:
```diff
@@ -3,18 +3,29 @@ import { store as noticesStore } from '@wordpress/notices';
  */
 export default function FlashMessage( {
-	key = 'updated',
+	id = DEFAULT_PARAM_NAME,
```
@millerf The reserved `key` prop warning was also fixed in #105912. Happy for it to be changed back to `id` again when this PR is rebased.

millerf replied:
Yes. I'll do that. Thanks for the heads up!

millerf replied:
Done!

millerf committed:
Use regular link for website URL

millerf committed:
Fix import order

millerf committed:
Let the component use enabled

millerf committed:
Update client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx

Co-authored-by: Philip Jackson <p-jackson@users.noreply.github.com>

millerf committed:
Fixing some lint errors and add translators

millerf committed:
Fix some lint error/some copies

millerf committed:
Return null

millerf committed:
Update me-monetize.ts

millerf committed:
Fix errors snackbars

millerf commented:
Updated the descriptions visuals to reflect latest changes

millerf committed:
Fix lint

paulopmt1 reviewed (commented):
Thanks for keeping iterating on this @millerf!

1. The auto-renew toggle is sending the correct data to the backend, but it's not updating after we use it, so after the first click, the subsequent usages of it will fail since the backend will return an error.

	https://github.com/user-attachments/assets/6f444385-63da-4a65-a637-db8ddd74317a

	Although we need to fix this toggle issue, if it helps you to do it on a follow-up, I'm okay with it.

2. The stop experience is returning an error. Can we also add a confirmation modal before executing it? This is important so we don't lose subscriptions unintentionally. I liked the fact we're redirecting after the deletion is completed, though!

	https://github.com/user-attachments/assets/9d5b4c97-77fe-4e61-a56a-336f9cd6f310



3. It's interesting to have that time of messaging after the user did the cancel action. Currently, it shows all the time, which feels weird as the user didn't say they want to leave us.

	<img width="797" height="629" alt="Image" src="https://github.com/user-attachments/assets/c8dab962-7f5d-4c6e-b30b-0cabb69d22f9" />

4. Is this breakline intentional? The previous one was inline, which feels better since it uses less space.

| Before | After |
|--------|--------|
| <img width="599" height="244" alt="Image" src="https://github.com/user-attachments/assets/f525add1-e97c-45f1-bd6b-0775fd5d1c5b" /> | <img width="655" height="368" alt="Image" src="https://github.com/user-attachments/assets/8dff0805-712a-4997-afa5-8b05653a239f" /> |

p-jackson reviewed (commented):
Apart from the small comment about periods at the end of lines, from the HD code point of view I think this is good now. I did leave another comment about the `monetizeSubscriptionQuery`, but I also don't want our continued conversation around how the `api-queries` package should work get in the way of merging this PR 🙂

fushar commented:
@ all reviewers: as noted in the Slack thread, I think in the general the functionality works. So it should be fine to fix CI in this PR, merge it, and then polish remaining issues in follow-up PRs. Let's not block this PR any longer 😄

millerf committed:
Fix subscription item link

millerf committed:
Fix subscription enable toggle

millerf committed:
Fix issue when cancelling a subscription

- Now pass a NavigateOptions to addFlashMessage
- Remove preemptively the cancelled subscription from the list

millerf commented:
@paulopmt1 Thanks for the review! 

> Is this breakline intentional? The previous one was inline, which feels better since it uses less space.

Nice catch. Fixed in 2eec3b6436f3f5d47b68d9e126ee6c6ebd610431

> The auto-renew toggle is sending the correct data to the backend, but it's not updating after we use it

Done in 0a4be81f3287ac589e7e9b9f6f1594d0445bb680. I was invalidating the query cache before setting the right data. I wasn't expecting the data to be fetched again right after the invalidation. This is news to me...

> t's interesting to have that time of messaging after the user did the cancel action. Currently, it shows all the time, which feels weird as the user didn't say they want to leave us.

Fixed in c6d7356fefaf2d25b47525b4e2f300e4dc769e8e. There was an issue with addFlashMessage that was doing a `new URL()` on a relative path. I had to modify the params from addFlashMessage and use a solution closer to Tanstack router paradigm.

> "We'll be sorry to see you go"

This is actually intended as @sirbrillig mentionned

millerf commented:
> I think this updated client/me/memberships/subscription.jsx but not client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx

@p-jackson Yes, you are right. It is fine, even if not really in the scope of this issue ;)

sirbrillig commented on client/dashboard/me/monetize-subscriptions/monetize-subscription.tsx:
```diff
@@ -0,0 +1,380 @@
+	const isRenewable = !! (
+		subscription &&
+		( subscription.renew_interval || subscription.is_renewable )
+	); // can remove renew_interval once backend is deployed
```
What does this comment mean? Is there a backend change pending? If so, can we put a link here to the Linear issue or the PR number?

millerf replied:
No, this is a comment that was already there there: https://github.com/Automattic/wp-calypso/blob/trunk/client/me/memberships/subscription.jsx#L45
This is the original one: https://github.com/Automattic/wp-calypso/commit/82b5114d1f307987daa5877586abf417dfeb3df9#diff-ba3dd79cc2302933f56ce3e57df1594680165aaf9fd17db991f9162b369f0b9aR34-R46

And I am not sure what it meant, as both parameters are returned (see in namespace.php#L1122 )

sirbrillig replied:
Hm. I'm guessing it's something like "once `is_renewable` is available, we can stop also checking `renew_interval`". Maybe `is_renewable` didn't exist when this was added.

millerf replied:
Fixed in 5f7b97534160339950d92511e73c62ec6a3223b9

sirbrillig committed:
Make notices end in periods.

Co-authored-by: Philip Jackson <p-jackson@users.noreply.github.com>

sirbrillig commented on .gitignore:
```diff
@@ -133,3 +133,6 @@ packages/plans-grid-next/static/mockServiceWorker.js
 output/
+
+# Claude settings
+CLAUDE.md
```
Since there's another PR for this change, I think it should be removed from this PR.

millerf replied:
True.

millerf replied:
Done!

sirbrillig reviewed (approved):
Other than the two comments I've left inline, I think this is good to go.

millerf committed:
Remove changes to gitignore

millerf committed:
Rely on subscription.is_renewable

millerf committed:
Fix tests

millerf committed:
Fix type issue

millerf committed:
Update index.tsx

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/25097326

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @millerf for including a screenshot in the description! This is really helpful for our translators.
